storage/innobase/xtrabackup/src/ds_decrypt.c

```diff
@@ -0,0 +1,845 @@
+		if (!gcrypt_version) {
+			msg("decrypt: failed to initialize libgcrypt\n");
+			return NULL;
+		} else {
+			msg("decrypt: using gcrypt %s\n", gcrypt_version);
+		}
+	}
+
+	/* Disable the gcry secure memory, not dealing with this for now */
+	gcry_error = gcry_control(GCRYCTL_DISABLE_SECMEM, 0);
+	if (gcry_error) {
+		msg("decrypt: unable to disable libgcrypt secmem - "
+		    "%s : %s\n",
+		    gcry_strsource(gcry_error),
+		    gcry_strerror(gcry_error));
+		return NULL;
+	}
+
+	/* Finalize gcry initialization. */
+	gcry_error = gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+	if (gcry_error) {
+		msg("decrypt: unable to finish libgcrypt initialization - "
+		    "%s : %s\n",
+		    gcry_strsource(gcry_error),
+		    gcry_strerror(gcry_error));
+		return NULL;
+	}
+
+	/* Set up the iv length */
+	encrypt_iv_len = gcry_cipher_get_algo_blklen(ds_decrypt_encrypt_algo);
+	xb_a(encrypt_iv_len > 0);
+
+	/* Now set up the key */
+	if (ds_decrypt_encrypt_key == NULL &&
+	    ds_decrypt_encrypt_key_file == NULL) {
+		msg("decrypt: no encryption key or key file specified.\n");
+		return NULL;
+	} else if (ds_decrypt_encrypt_key && ds_decrypt_encrypt_key_file) {
+		msg("decrypt: both encryption key and key file specified.\n");
+		return NULL;
+	} else if (ds_decrypt_encrypt_key_file) {
+		if (!xb_crypt_read_key_file(ds_decrypt_encrypt_key_file,
+					    (void**)&ds_decrypt_encrypt_key,
+					    &encrypt_key_len)) {
+			msg("decrypt: unable to read encryption key file"
+			    " \"%s\".\n", ds_decrypt_encrypt_key_file);
+			return NULL;
+		}
+	} else if (ds_decrypt_encrypt_key) {
+		encrypt_key_len = strlen(ds_decrypt_encrypt_key);
+	} else {
+		msg("decrypt: no encryption key or key file specified.\n");
+		return NULL;
+	}
+
+	/* Create and initialize the worker threads */
+	threads = create_worker_threads(ds_decrypt_encrypt_threads);
+	if (threads == NULL) {
+		msg("decrypt: failed to create worker threads.\n");
+		return NULL;
+	}
+
+	ctxt = (ds_ctxt_t *) my_malloc(sizeof(ds_ctxt_t) +
+				       sizeof(ds_decrypt_ctxt_t),
+				       MYF(MY_FAE));
+
+	decrypt_ctxt = (ds_decrypt_ctxt_t *) (ctxt + 1);
+	decrypt_ctxt->threads = threads;
+	decrypt_ctxt->nthreads = ds_decrypt_encrypt_threads;
+
+	ctxt->ptr = decrypt_ctxt;
+	ctxt->root = my_strdup(root, MYF(MY_FAE));
+
+	return ctxt;
+}
+
+static
+ds_file_t *
+decrypt_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat)
+{
+	ds_ctxt_t		*dest_ctxt;
+
+	ds_decrypt_ctxt_t	*crypt_ctxt;
+	ds_decrypt_file_t	*crypt_file;
+
+	char			new_name[FN_REFLEN];
+	ds_file_t		*file;
+
+	xb_ad(ctxt->pipe_ctxt != NULL);
+	dest_ctxt = ctxt->pipe_ctxt;
+
+	crypt_ctxt = (ds_decrypt_ctxt_t *) ctxt->ptr;
+
+
+	file = (ds_file_t *) my_malloc(sizeof(ds_file_t) +
+				       sizeof(ds_decrypt_file_t),
+				       MYF(MY_FAE|MY_ZEROFILL));
+
+	crypt_file = (ds_decrypt_file_t *) (file + 1);
+
+	/* Remove the .xbcrypt extension from the filename */
+	strcpy(new_name, path);
```

Comment:
Code is written with assumption that path is no longer than FN_REFLEN. However I agree that strncpy is safer choice.